cpp/src/gandiva/gdv_function_stubs.cc

```diff
@@ -794,6 +795,88 @@ const char* gdv_fn_initcap_utf8(int64_t context, const char* data, int32_t data_
   *out_len = out_idx;
   return out;
 }
+
+GANDIVA_EXPORT
+const char* gdv_fn_mask_first_n(int64_t context, const char* data, int32_t data_len,
+                                int32_t n_to_mask, int32_t* out_len) {
+  if (data_len <= 0) {
+    *out_len = 0;
+    return nullptr;
+  }
+
+  if (n_to_mask < 0) {
+    n_to_mask = n_to_mask * (-1);
+  }
+
+  *out_len = data_len;
+
+  char* out = reinterpret_cast<char*>(gdv_fn_context_arena_malloc(context, *out_len));
+  if (out == nullptr) {
+    gdv_fn_context_set_error_msg(context, "Could not allocate memory for output string");
+    *out_len = 0;
+    return nullptr;
+  }
+
+  // do the masking
+  for (int i = 0; i < data_len; ++i) {
+    if(isdigit(data[i]) && i < n_to_mask) {
```

Comment:
This is going to be slow... Please construct a static array as follows:
final char[256] mask_array = {
 (char)0, (char)1, ....
 'n', 'n', // for the ascii values of '0' to '9'
 'X', 'X', // for the ascii values of 'A' to 'Z'
 'x', 'x', // for the ascii values of 'a' to 'z'
};

With this array, this loop becomes:
out[i] = mask_array[data[i]];

Reply:
Can you check how Hive implements this for utf-8 strings? This logic works for ascii input, it doesn't work for utf-8 input

Reply:
In hive implementation it does a similar logic of what I use, as it can bee seen here: https://github.com/apache/hive/blob/c8c5d2bee7f357ab873d0f3366ae94624d9a73fe/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFMask.java#L240. Should I use your static array, or leave this as it is?

Reply:
Lets use a static array - this is going to be more performant for a little extra memory

Reply:
With utf-8, a character can be multiple bytes. Masking first n will require masking first n characters - that will change the logic completely. With multi-byte chars, this logic is not going to work. Does Hive also assume ascii characters?

Reply:
Hive uses the class Character wich is based upon the unicode characters with 16 bits (2bytes)

Reply:
To mask first 2 characters, is it masking 4 bytes in total?

Reply:
Added the static mask_array